Userland/Libraries/LibGUI/VimEditingEngine.cpp

```diff
@@ -63,8 +63,111 @@ bool VimEditingEngine::on_key_in_insert_mode(const KeyEvent& event)
     return false;
 }
 
+char VimEditingEngine::numeric_key_value(const KeyCode key)
+{
+    switch (key) {
+    case (KeyCode::Key_0):
+        return '0';
+    case (KeyCode::Key_1):
+        return '1';
+    case (KeyCode::Key_2):
+        return '2';
+    case (KeyCode::Key_3):
+        return '3';
+    case (KeyCode::Key_4):
+        return '4';
+    case (KeyCode::Key_5):
+        return '5';
+    case (KeyCode::Key_6):
+        return '6';
+    case (KeyCode::Key_7):
+        return '7';
+    case (KeyCode::Key_8):
+        return '8';
+    case (KeyCode::Key_9):
+        return '9';
+    default:
+        return '\0';
+    }
+}
+
+void VimEditingEngine::clear_command_multiplier()
+{
+    m_command_multiplier.clear();
+}
+
+void VimEditingEngine::vim_move_down(const KeyEvent& event)
+{
+    // FIXME: probably not an efficient way to do this
+    if (m_command_multiplier.is_empty()) {
+        move_one_down(event);
+    } else {
+        String multiplier_string = m_command_multiplier.build();
+        uint32_t multiplier = multiplier_string.to_uint().value();
+        for (uint32_t i = 0; i < multiplier; ++i) {
+            move_one_down(event);
+        }
+        clear_command_multiplier();
+    }
+}
+
+void VimEditingEngine::vim_move_up(const KeyEvent& event)
+{
+    if (m_command_multiplier.is_empty()) {
+        move_one_up(event);
+    } else {
+        String multiplier_string = m_command_multiplier.build();
+        uint32_t multiplier = multiplier_string.to_uint().value();
+        for (uint32_t i = 0; i < multiplier; ++i) {
+            move_one_up(event);
+        }
+        clear_command_multiplier();
+    }
+}
```

Comment:
this logic shouldn't be repeated this many times. I would make a helper function that takes a lambda expression and evaluates it `m_command_multiplier` times, then resets `m_command_multiplier`. then I would wrap the actions in `on_key_in_normal_mode`

Reply:
Super! I'll have a look at how lambdas are implemented in C++ and fix that :)